src/intptrcast.rs

```diff
@@ -107,9 +106,10 @@ impl<'mir, 'tcx> GlobalState {
                     slack,
                 );
 
-                // Remember next base address.  If this allocation is zero-sized, leave a gap
-                // of at least 1 to avoid two allocations having the same base address.
-                global_state.next_base_addr = base_addr.checked_add(max(size.bytes(), 1)).unwrap();
+                // Remember next base address.  Leave a gap of at least 1 to avoid two zero-sized allocations
+                // having the same base address, and to avoid ambiguous provenance for the address between two
+                // allocations.
+                global_state.next_base_addr = base_addr.checked_add(size.bytes() + 1).unwrap();
```

Comment:
This should use `checked_add`, not `+`.

Reply:
I doubt that `size.bytes()` can actually be `usize::MAX`, but it doesn't hurt to be careful

Reply:
Yeah, you are probably right but I consider unchecked arithmetic to be essentially banned from this repo. ;)